File: src/test/java/dev/nytweetdeck/post/TranslationMemoryTest.java
package dev.nytweetdeck.post;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

class TranslationMemoryTest {
    private TranslationMemory.Key key(String id) {
        return new TranslationMemory.Key("account", "note", id, "en", "ja", "original");
    }

    @Test
    void keepsOnlySuccessAndEvictsLeastRecentlyUsed() {
        var memory = new TranslationMemory<String>(2);
        memory.load(key("1"), () -> "one", value -> true);
        memory.load(key("2"), () -> "two", value -> true);
        assertThat(memory.load(key("1"), () -> "wrong", value -> true)).isEqualTo("one");
        memory.load(key("3"), () -> "three", value -> true);
        assertThat(memory.load(key("2"), () -> "new two", value -> true)).isEqualTo("new two");
        memory.load(key("missing"), () -> "unavailable", value -> false);
        assertThat(memory.load(key("missing"), () -> "available", value -> true)).isEqualTo("available");
        assertThatThrownBy(() -> memory.load(key("failed"), () -> { throw new IllegalStateException("failure"); }, value -> true))
                .hasMessage("failure");
        assertThat(memory.load(key("failed"), () -> "recovered", value -> true)).isEqualTo("recovered");
        assertThat(memory.size()).isEqualTo(2);
    }

    @Test
    void sharesPendingWithoutHoldingTheLockDuringUpstreamWork() throws Exception {
        var memory = new TranslationMemory<String>(1);
        var started = new CountDownLatch(1);
        var release = new CountDownLatch(1);
        var joined = new CountDownLatch(1);
        var calls = new AtomicInteger();
        var first = CompletableFuture.supplyAsync(() -> memory.load(key("1"), () -> {
            calls.incrementAndGet();
            started.countDown();
            try { if (!release.await(3, TimeUnit.SECONDS)) throw new IllegalStateException("timeout"); }
            catch (InterruptedException e) { throw new IllegalStateException(e); }
            return "translated";
        }, value -> true));
        assertThat(started.await(3, TimeUnit.SECONDS)).isTrue();
        var second = CompletableFuture.supplyAsync(() -> memory.load(key("1"), () -> "wrong", value -> true, () -> {}, joined::countDown));
        try {
            assertThat(joined.await(3, TimeUnit.SECONDS)).isTrue();
            assertThat(memory.load(key("2"), () -> "other", value -> true)).isEqualTo("other");
            assertThat(memory.pendingCount()).isEqualTo(1);
        } finally { release.countDown(); }
        assertThat(first.get(3, TimeUnit.SECONDS)).isEqualTo("translated");
        assertThat(second.get(3, TimeUnit.SECONDS)).isEqualTo("translated");
        assertThat(calls.get()).isEqualTo(1);
    }
}

File: android/app/src/test/java/dev/nytweetdeck/android/data/TranslationMemoryTest.kt
package dev.nytweetdeck.android.data

import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import org.junit.Assert.*
import org.junit.Test

class TranslationMemoryTest {
    @Test fun retainsSuccessEvictsLeastRecentlyUsedAndSeparatesKeys() {
        val memory = TranslationMemory<Triple<String, String, String>, String>(2)
        val japanese = Triple("account", "55", "ja")
        val french = Triple("account", "55", "fr")
        val otherAccount = Triple("other", "55", "ja")
        assertEquals("Japanese", memory.getOrLoad(japanese) { "Japanese" })
        memory.getOrLoad(french) { "French" }
        assertEquals("Japanese", memory.getOrLoad(japanese) { error("cached") })
        memory.getOrLoad(otherAccount) { "Other" }
        assertEquals("Japanese", memory.getOrLoad(japanese) { error("cached") })
        assertEquals("Reload", memory.getOrLoad(french) { "Reload" })
        assertEquals(2, memory.size)
    }

    @Test fun unavailableAndFailuresCanBeRetried() {
        val memory = TranslationMemory<String, String?>(2)
        assertNull(memory.getOrLoad("note", cacheable = { it != null }) { null })
        assertEquals(0, memory.size)
        assertThrows(IllegalStateException::class.java) {
            memory.getOrLoad("note") { error("temporary") }
        }
        assertEquals(0, memory.inFlight)
        assertEquals("Translated", memory.getOrLoad("note") { "Translated" })
    }

    @Test fun concurrentReadersSharePendingResultAndLaterReadersReuseIt() {
        val memory = TranslationMemory<String, String>(2)
        val executor = Executors.newFixedThreadPool(2)
        val started = CountDownLatch(1)
        val joined = CountDownLatch(1)
        val release = CountDownLatch(1)
        val calls = AtomicInteger()
        try {
            val first = executor.submit<String> {
                memory.getOrLoad("same") {
                    calls.incrementAndGet()
                    started.countDown()
                    check(release.await(5, TimeUnit.SECONDS))
                    "Translated"
                }
            }
            assertTrue(started.await(5, TimeUnit.SECONDS))
            val second = executor.submit<String> {
                memory.getOrLoad("same", onJoin = { joined.countDown() }) { error("duplicate") }
            }
            assertTrue(joined.await(5, TimeUnit.SECONDS))
            release.countDown()
            assertEquals("Translated", first.get(5, TimeUnit.SECONDS))
            assertEquals("Translated", second.get(5, TimeUnit.SECONDS))
            assertEquals("Translated", memory.getOrLoad("same") { error("cached") })
            assertEquals(1, calls.get())
            assertEquals(0, memory.inFlight)
        } finally {
            release.countDown()
            executor.shutdownNow()
        }
    }
}
